src/Autofac/Core/Registration/DefaultRegisteredServicesTracker.cs

```diff
@@ -199,42 +188,33 @@ public bool IsRegistered(Service service)
         {
             if (service == null) throw new ArgumentNullException(nameof(service));
 
-            lock (_synchRoot)
-            {
-                return GetInitializedServiceInfo(service).IsRegistered;
-            }
+            return GetInitializedServiceInfo(service).IsRegistered;
         }
 
         /// <inheritdoc />
         public IEnumerable<IComponentRegistration> RegistrationsFor(Service service)
         {
             if (service == null) throw new ArgumentNullException(nameof(service));
 
-            lock (_synchRoot)
-            {
-                var info = GetInitializedServiceInfo(service);
-                return info.Implementations.ToList();
-            }
+            var info = GetInitializedServiceInfo(service);
+            return info.Implementations.ToList();
```

Comment:
I'm thinking about how `Implementations` is accessed here and in a couple of other spots. Ostensibly once the `ServiceRegistrationInfo` is initialized, that's a fixed list, right? Could `Implementations` be cached as a generated list and avoid the allocation during `RegistrationsFor`?

I may not be totally following things through, so I could be wrong.

Reply:
Also, it's OK if that's a separate PR. It doesn't have to block this one.

Reply:
Possibly, although the `RegistrationsFor` method is now not really used at runtime (in favour of `ServiceRegistrationsFor`).

I can take a look at the caching of the ServiceRegistration list. but perhaps do that in a separate PR, because it would be another set of fairly substantial changes unrelated to concurrency.

Reply:
Sounds good!